compiler/optimizer/AsyncCheckInsertion.cpp

```diff
@@ -102,7 +119,18 @@ bool TR_AsyncCheckInsertion::shouldPerform()
    if (comp()->isProfilingCompilation() || comp()->generateArraylets())
       return false;
 
-   if (comp()->getOption(TR_EnableOSR))  // cannot move async checks around arbitrarily if OSR is possible
+   // It is not safe to add an asynccheck under involuntary OSR
+   // as a transition may have to occur at the added point and the
+   // required infrastructure may not exist
+   //
+   if (comp()->getOption(TR_EnableOSR) && comp()->getOSRMode() == TR::involuntaryOSR)
+      return false;
+
+   // Should not be performed at very hot or scorching. Sampling does
+   // not matter at these points, as recompilation will not occur based
+   // on it.
+   //
+   if (comp()->getMethodHotness() >= veryHot)
```

Comment:
Thanks. Yes I would support using the more precise condition as in the commit that you linked to.